In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import yfinance as yf
import pandas as pd

In [ ]:
tickers = ['ORCL', 'GLD', 'XOM', 'KO']

prices = yf.download(tickers,start='2022-01-01', end='2025-09-04')['Close']

daily_rend = prices.pct_change().dropna()
rend = daily_rend.mean()

/tmp/ipython-input-402279560.py:3: FutureWarning: YF.download() has changed argument auto_adjust default to True
  prices = yf.download(tickers,start='2022-01-01', end='2025-09-04')['Close']
[*********************100%***********************]  4 of 4 completed


In [ ]:
# Matriz de pesos y Matriz de Varianza-Covarianza
w = np.array([0.25, 0.25, 0.25, 0.25])

cov = daily_rend.cov()

In [ ]:
rend_p = float(np.dot(rend, w))
rend_diario = rend_p * 100
rend_anual = rend_p * 100 * 252

df_rend = pd.DataFrame({
    "Métrica": ["Rendimiento Diario (%)", "Rendimiento Anual (%)"],
    "Valor": [rend_diario, rend_anual]
})

df_rend

Métrica      Valor
0  Rendimiento Diario (%)   0.083102
1   Rendimiento Anual (%)  20.941769

In [ ]:
var_p = float(w.T @ cov @ w)
var_diaria = var_p * 100
var_anual = var_p * 252 * 100

df_var = pd.DataFrame({
    "Métrica": [
        "Varianza Diaria (%)",
        "Varianza Anual (%)"
    ],
    "Valor": [var_diaria, var_anual]
})

df_var

Métrica     Valor
0  Varianza Diaria (%)  0.007774
1   Varianza Anual (%)  1.958995

In [ ]:
vol_diaria = np.sqrt(var_p) * 100
vol_anual = np.sqrt(var_p * 252) * 100

df_vol = pd.DataFrame({
    "Métrica": ["Volatilidad Diaria (%)", "Volatilidad Anual (%)"],
    "Valor": [vol_diaria, vol_anual]
})

df_vol

Métrica      Valor
0  Volatilidad Diaria (%)   0.881691
1   Volatilidad Anual (%)  13.996412

In [ ]:
# Coeficiente de Variación
cv = vol_anual/rend_anual
cv

np.float64(0.6683490659765637)

In [ ]:
# Matriz de Correlación
cor = daily_rend.corr()
cor

Ticker       GLD        KO      ORCL       XOM
Ticker                                        
GLD     1.000000  0.129559  0.064217  0.180314
KO      0.129559  1.000000  0.081168  0.160407
ORCL    0.064217  0.081168  1.000000  0.093233
XOM     0.180314  0.160407  0.093233  1.000000

In [ ]:
suma_inferior = np.tril(cor, k=-1).sum()
suma_inferior

np.float64(0.7088985449880136)

# Conclusión


La matriz de correlación muestra que los activos considerados en el portafolio (GLD, KO, ORCL y XOM) se mueven en la misma dirección, pero con relaciones muy bajas entre sí, ninguna superior a 0.20. Esto es positivo porque significa que no reaccionan de la misma manera ante los cambios del mercado, lo que se traduce en una mayor diversificación. Dentro de este grupo, el oro (GLD) es el activo que menos se relaciona con los demás, lo que lo convierte en un buen refugio en escenarios de incertidumbre. En general, contar con correlaciones bajas es una señal de que la combinación de activos está bien pensada, pues se evita que todos sufran pérdidas al mismo tiempo.


En cuanto al riesgo, cada activo individual arrastra su propio riesgo específico, ligado a factores particulares de su industria: la tecnología en Oracle, el consumo en Coca-Cola, la energía en Exxon o los commodities en el caso del oro. Estos riesgos, considerados de manera aislada, podrían impactar fuertemente a un inversionista si solo tuviera uno de estos activos en cartera. Sin embargo, al juntarlos en un portafolio equilibrado, dichos riesgos se compensan unos con otros y se reduce notablemente el riesgo no sistemático. Esto significa que el portafolio aprovecha la diversificación para eliminar gran parte de la incertidumbre propia de cada empresa o sector, quedando principalmente el riesgo general del mercado, que es inevitable y que está asociado al contexto económico global.


Por último, al revisar las métricas de rendimiento y riesgo del portafolio, los resultados son favorables. El rendimiento anual esperado ronda el 21%, lo cual resulta bastante atractivo frente a la media del mercado y a lo que ofrecerían los activos por separado. En cuanto al riesgo, la volatilidad anual se ubica en aproximadamente 14%, un nivel manejable considerando la rentabilidad proyectada, mientras que la varianza anual es cercana al 2%, lo que confirma que la dispersión de los rendimientos no es excesiva. Además, al calcular el coeficiente de variación se observa un valor menor a 1, lo que indica que el rendimiento compensa adecuadamente el riesgo asumido. En otras palabras, por cada unidad de riesgo tomada, el portafolio ofrece un rendimiento proporcionalmente alto. Esta combinación refleja una relación riesgo-rendimiento sólida y justifica la composición del portafolio. Gracias a que los activos presentan correlaciones bajas, la diversificación no solo funciona en teoría, sino que se refleja en la práctica al mantener el riesgo en niveles controlados y al mismo tiempo alcanzar un rendimiento competitivo. En conclusión, el portafolio cumple con el objetivo de balancear rentabilidad y seguridad, mostrando que una adecuada selección y combinación de activos puede mejorar significativamente el perfil de riesgo de un inversionista.


